### PR TITLE
Login

### PR DESCRIPTION
**Pull request by HVlad1**
Created Login screen

### CODE REVIEW

**Comment by MunBeniamin** [comment]
Add more team members as reviewers please, and the base should be develop not master.

File: Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+    
+    @IBOutlet weak var passwordLabel: UILabel!
+    @IBOutlet weak var emailLabel: UILabel!
+    @IBOutlet weak var logInLabel: UILabel!
+    @IBOutlet weak var emailTextField: UITextField!
+    @IBOutlet weak var passwordTextField: UITextField!
+    @IBOutlet weak var loginButton: UIButton!
+    @IBOutlet weak var notRegisteredLabel: UILabel!
+    @IBOutlet weak var signupButton: UIButton!
+    @IBOutlet weak var forgotPasswordButton: UIButton!
```
Comment (IgnatCamelia) [inline Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift]: Please use private outlets

File: Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    
+    let colorText = UIColor(red: 37/255, green: 22/255, blue: 5/255, alpha: 1)
+    let colorView = UIColor(red: 124/255, green: 122/255, blue: 122/255, alpha: 1)
```
Comment (IgnatCamelia) [inline Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift]: Make a UIColor extension so you won't need to normalise the colors everytime.

File: Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        ])
+        
+    }
+    
+
+    @IBAction func loginTapped(_ sender: Any) {
+        let userEmail = emailTextField.text
+        let userPassword = passwordTextField.text
+        
+        if(userEmail == "" || userPassword == "") {
```
Comment (IgnatCamelia) [inline Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift]: Space after if, no need to use ()

File: Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+            alertMessage(userMessage: "All fields are required")
+        } else {
+            let mail = isValidEmail(emailID: userEmail!)
```
Comment (IgnatCamelia) [inline Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift]: Do not use force unwrap

File: Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+            if(mail == false) {
```
Comment (IgnatCamelia) [inline Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift]: Space after if, no need to use ()

File: Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+                alertMessage(userMessage: "Please enter a valid email addres!")
+                emailLabel.textColor = .red
+            } else {
+                if(isValidPassword(password: userPassword!) == false ) {
```
Comment (IgnatCamelia) [inline Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift]: Space after if, no need to use (). Do not use force unwrap.

File: Internship2022TeamPatri/Internship2022TeamPatri/Views/Auth/Login/LogIn.storyboard [hunk 7]
```diff
@@ -0,0 +1,147 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<document type="com.apple.InterfaceBuilder3.CocoaTouch.Storyboard.XIB" version="3.0" toolsVersion="19455" targetRuntime="iOS.CocoaTouch" propertyAccessControl="none" useAutolayout="YES" useTraitCollections="YES" useSafeAreas="YES" colorMatched="YES" initialViewController="3t1-SN-qvL">
+    <device id="retina6_7" orientation="portrait" appearance="light"/>
+    <dependencies>
+        <plugIn identifier="com.apple.InterfaceBuilder.IBCocoaTouchPlugin" version="19454"/>
```
Comment (IgnatCamelia) [inline Internship2022TeamPatri/Internship2022TeamPatri/Views/Auth/Login/LogIn.storyboard:5]: Do we really need to use a storyboard for Login?

File: Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+    //Alert
+    func alertMessage(userMessage: String) {
+        let myAllert = UIAlertController(title: "Something went wrong", message: userMessage, preferredStyle: UIAlertController.Style.alert)
+        let ok = UIAlertAction(title: "OK", style: UIAlertAction.Style.default, handler:  nil)
+        myAllert.addAction(ok)
+        present(myAllert, animated: true, completion: nil)
+    }
+   
+class Utilities {
+
```
Comment (IgnatCamelia) [inline Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift:125]: Remove space.

File: Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+        let ok = UIAlertAction(title: "OK", style: UIAlertAction.Style.default, handler:  nil)
+        myAllert.addAction(ok)
+        present(myAllert, animated: true, completion: nil)
+    }
+   
+class Utilities {
+    
+    //Style the text fields
+    static func styleTextField(textfield:UITextField) {
+
```
Comment (IgnatCamelia) [inline Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift:127]: Remove space.

File: Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+    // Adding shadow to the button
+    static func styleShadowButton(button: UIButton){
+        button.layer.shadowColor = UIColor(red: 0, green: 0, blue: 0, alpha: 0.25).cgColor
+        button.layer.shadowOffset = CGSize(width: 0.0, height: 3.0)
+        button.layer.shadowOpacity = 1.0
+        button.layer.shadowRadius = 0.0
+        button.layer.masksToBounds = false
+    }
+}
+
```
Comment (IgnatCamelia) [inline Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift]: Remove space.

File: Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+                    emailLabel.textColor = colorText
+                    passwordLabel.textColor = .red
+                } else {
+                    alertMessage(userMessage: "Success!")
+                    passwordLabel.textColor = colorText
+                }
+            }
+        }
+    }
+
```
Comment (IgnatCamelia) [inline Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift:100]: Remove space.

File: Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+            .foregroundColor: UIColor.black,
+            .font: UIFont.boldSystemFont(ofSize: 16.0)
+        ])
+        passwordTextField.attributedPlaceholder = NSAttributedString(string: "", attributes: [
+            .foregroundColor: UIColor.black,
+            .font: UIFont.boldSystemFont(ofSize: 16.0)
+        ])
+        
+    }
+
```
Comment (IgnatCamelia) [inline Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift:76]: Remove space.

File: Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+              string: "Forgot password?",
+              attributes: yourAttributes
+           )
+        signupButton.setAttributedTitle(attributeString, for: .normal)
+        forgotPasswordButton.setAttributedTitle(attributeStringForgotPassword, for: .normal)
+        
+        setUpElements()
+    }
+    
+
```
Comment (IgnatCamelia) [inline Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift]: To many spaces

File: Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+           )
+        signupButton.setAttributedTitle(attributeString, for: .normal)
+        forgotPasswordButton.setAttributedTitle(attributeStringForgotPassword, for: .normal)
+        
+        setUpElements()
+    }
+    
+    
+    
+    func setUpElements() {
```
Comment (IgnatCamelia) [inline Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift]: This can be private

File: Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+        ])
+        passwordTextField.attributedPlaceholder = NSAttributedString(string: "", attributes: [
+            .foregroundColor: UIColor.black,
+            .font: UIFont.boldSystemFont(ofSize: 16.0)
+        ])
+        
+    }
+    
+
+    @IBAction func loginTapped(_ sender: Any) {
```
Comment (IgnatCamelia) [inline Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift]: Private

File: Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+                } else {
+                    alertMessage(userMessage: "Success!")
+                    passwordLabel.textColor = colorText
+                }
+            }
+        }
+    }
+    
+    
+    @IBAction func signupTapped(_ sender: Any) {
```
Comment (IgnatCamelia) [inline Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift]: Private

File: Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+    }
+
+    //Email validation
+    func isValidEmail(emailID: String) ->Bool {
```
Comment (IgnatCamelia) [inline Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift]: Private

File: Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift [hunk 18]
```diff
@@ -0,0 +1,169 @@
...
+        let emailRegEx = "[A-Z0-9a-z._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,64}"
+        let emailTest = NSPredicate(format: "SELF MATCHES %@", emailRegEx)
+        return emailTest.evaluate(with: emailID)
+    }
+    
+    //Password validation
+    func isValidPassword(password: String) ->Bool{
```
Comment (IgnatCamelia) [inline Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift]: Private

File: Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift [hunk 19]
```diff
@@ -0,0 +1,169 @@
...
+        let passwordReg = "^(?=.*[a-z])(?=.*[A-Z])(?=.*\\d)[a-zA-Z\\d]{8,}$"
+        let passwordTesting = NSPredicate(format: "SELF MATCHES %@", passwordReg)
+        return passwordTesting.evaluate(with: password)
+    }
+    
+    //Alert
+    func alertMessage(userMessage: String) {
```
Comment (IgnatCamelia) [inline Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift]: Private

File: Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift (hunk 2 above)
Comment (MunBeniamin) [reply Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift]: I would suggest to create color assets and create an enum to access each color.

File: Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift [hunk 20]
```diff
@@ -0,0 +1,169 @@
...
+        
+        // Remove border
+        textfield.borderStyle = .none
+        
+        // Add the line to the text field
+        textfield.layer.addSublayer(bottomLine)
+    }
+    
+    //Style the Log in Button
+    static func styleLogInButton(button: UIButton) {
```
Comment (MunBeniamin) [inline Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift]: Why are these marked as "static"?

File: Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift [hunk 21]
```diff
@@ -0,0 +1,163 @@
...
+                    passwordLabel.textColor = .red
+                } else {
+                    alertMessage(userMessage: "Success!")
+                    passwordLabel.textColor = .colorText
+                }
+            }
+        }
+    }
+    
+    @IBAction func signupTapped(_ sender: Any) {
```
Comment (IgnatCamelia) [inline Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift]: Private action.

**Review by IgnatCamelia — APPROVED** [review]

File: Internship2022TeamPatri/Internship2022TeamPatri/Base.lproj/Main.storyboard [hunk 22]
```diff
@@ -1,44 +1,164 @@
...
+                <placeholder placeholderIdentifier="IBFirstResponder" id="DXb-JU-xtl" userLabel="First Responder" customClass="UIResponder" sceneMemberID="firstResponder"/>
+            </objects>
+            <point key="canvasLocation" x="34.782608695652179" y="97.767857142857139"/>
         </scene>
     </scenes>
-    <resources>
-        <systemColor name="systemBackgroundColor">
-            <color white="1" alpha="1" colorSpace="custom" customColorSpace="genericGamma22GrayColorSpace"/>
-        </systemColor>
-    </resources>
```
Comment (patriciaghitun) [inline Internship2022TeamPatri/Internship2022TeamPatri/Base.lproj/Main.storyboard:43]: Please remove your previous work. This file (Main.storyboard) should be untouched since you moved your work to another file (LogIn.storyboard)

File: Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift [hunk 23]
```diff
@@ -0,0 +1,163 @@
...
+    @IBOutlet weak private var emailLabel: UILabel!
+    @IBOutlet weak private var logInLabel: UILabel!
+    @IBOutlet weak private var emailTextField: UITextField!
+    @IBOutlet weak private var passwordTextField: UITextField!
+    @IBOutlet weak private var loginButton: UIButton!
+    @IBOutlet weak private var notRegisteredLabel: UILabel!
+    @IBOutlet weak private var signupButton: UIButton!
+    @IBOutlet weak private var forgotPasswordButton: UIButton!
+    
+    let yourAttributesSignUp: [NSAttributedString.Key: Any] = [
```
Comment (patriciaghitun) [inline Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift]: Find a better name for “yourAttributes/yourAttributesSignUp”. You must give suggestive names to the variables so that each and every one of us should be able to figure out what that variables is doing without even looking at its type and its context. We have to know what a variable is for after checking its name.

File: Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift [hunk 24]
```diff
@@ -0,0 +1,163 @@
...
+    }
+    
+    @IBAction private func loginTapped(_ sender: Any) {
+        let userEmail = emailTextField.text
+        let userPassword = passwordTextField.text
+        
+        if userEmail == "" || userPassword == "" {
+            alertMessage(userMessage: "All fields are required")
+        } else {
+            let mail = isValidEmail(emailID: userEmail)
```
Comment (patriciaghitun) [inline Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift]: No need for "ID". Be consistent when naming something. If you chose isValidPassword(password:) then isValidEmail(email:) should be a better fit

File: Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift [hunk 25]
```diff
@@ -0,0 +1,163 @@
...
+    // Adding shadow to the button
+    final class func styleShadowButton(button: UIButton) {
+        button.layer.shadowColor = UIColor(red: 0, green: 0, blue: 0, alpha: 0.25).cgColor
+        button.layer.shadowOffset = CGSize(width: 0.0, height: 3.0)
+        button.layer.shadowOpacity = 1.0
+        button.layer.shadowRadius = 0.0
+        button.layer.masksToBounds = false
+    }
+}
+}
```
Comment (patriciaghitun) [inline Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift]: Please format your code: ctrl + i before commit

File: Internship2022TeamPatri/Internship2022TeamPatri/Views/Auth/Login/LogIn.storyboard (hunk 7 above)
Comment (patriciaghitun) [reply Internship2022TeamPatri/Internship2022TeamPatri/Views/Auth/Login/LogIn.storyboard:5]: I completely agree with Camelia, storyboards should be used for more complex flows. Instead, you could try xib / programmatic code

File: Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift [hunk 26]
```diff
@@ -0,0 +1,163 @@
...
+            .foregroundColor: UIColor.black,
+            .font: UIFont.boldSystemFont(ofSize: 16.0)
+        ])
+    }
+    
+    @IBAction private func loginTapped(_ sender: Any) {
+        let userEmail = emailTextField.text
+        let userPassword = passwordTextField.text
+        
+        if userEmail == "" || userPassword == "" {
```
Comment (patriciaghitun) [inline Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift]: instead of == “” check if field isEmpty

File: Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift [hunk 27]
```diff
@@ -0,0 +1,163 @@
...
+        let emailRegEx = "[A-Z0-9a-z._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,64}"
+        let emailTest = NSPredicate(format: "SELF MATCHES %@", emailRegEx)
+        return emailTest.evaluate(with: emailID)
+    }
+    
+    //Password validation
+    private func isValidPassword(password: String?) ->Bool{
+        let passwordReg = "^(?=.*[a-z])(?=.*[A-Z])(?=.*\\d)[a-zA-Z\\d]{8,}$"
+        let passwordTesting = NSPredicate(format: "SELF MATCHES %@", passwordReg)
+        return passwordTesting.evaluate(with: password)
```
Comment (patriciaghitun) [inline Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift]: Please be consistent about variable names. passwordTesting and emailTest

File: Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift [hunk 28]
```diff
@@ -0,0 +1,163 @@
...
+    private func alertMessage(userMessage: String) {
+        let myAllert = UIAlertController(title: "Something went wrong", message: userMessage, preferredStyle: UIAlertController.Style.alert)
+        let ok = UIAlertAction(title: "OK", style: UIAlertAction.Style.default, handler:  nil)
+        myAllert.addAction(ok)
+        present(myAllert, animated: true, completion: nil)
+    }
+   
+    class Utilities {
+    //Style the text fields
+    final class func styleTextField(textfield:UITextField) {
```
Comment (patriciaghitun) [inline Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift]: Remove unnecessary comments (style the text fields, that can be understood from the method's name)

File: Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift [hunk 29]
```diff
@@ -0,0 +1,163 @@
...
+    
+    @IBOutlet weak private var passwordLabel: UILabel!
+    @IBOutlet weak private var emailLabel: UILabel!
+    @IBOutlet weak private var logInLabel: UILabel!
+    @IBOutlet weak private var emailTextField: UITextField!
+    @IBOutlet weak private var passwordTextField: UITextField!
+    @IBOutlet weak private var loginButton: UIButton!
+    @IBOutlet weak private var notRegisteredLabel: UILabel!
+    @IBOutlet weak private var signupButton: UIButton!
+    @IBOutlet weak private var forgotPasswordButton: UIButton!
```
Comment (patriciaghitun) [inline Internship2022TeamPatri/Internship2022TeamPatri/Controllers/Auth/Login/LogInViewController.swift:21]: Please be consistent when naming your labels/variables/etc (logInLabel, loginButton). Please check the other variables as well

File: Internship2022TeamPatri/Internship2022TeamPatri/Base.lproj/LaunchScreen.storyboard [hunk 30]
```diff
@@ -11,15 +13,20 @@
             <objects>
                 <viewController id="01J-lp-oVM" sceneMemberID="viewController">
```
Comment (patriciaghitun) [inline Internship2022TeamPatri/Internship2022TeamPatri/Base.lproj/LaunchScreen.storyboard:14]: Please commit only your work, no need to add unnecessary changes to your commits

**Review by patriciaghitun — CHANGES_REQUESTED** [review]

**Review by patriciaghitun — APPROVED** [review]

**Review by MuresanCoralia — APPROVED** [review]
Perf ✨
